spikeextractors/extractors/nwbextractors/nwbextractors.py

```diff
@@ -318,28 +320,20 @@ def add_devices(recording: se.RecordingExtractor, nwbfile=None,
 
         if metadata is None:
             metadata = dict()
+            metadata['Ecephys'] = {}
+            metadata['Ecephys']['Device'] = []
 
-        if len(metadata.keys()) > 0:
-            if 'Ecephys' in metadata \
-                    and 'Device' in metadata['Ecephys'] \
-                    and type(metadata['Ecephys']['Device']) is list \
-                    and metadata['Ecephys']['Device']:
-                for j, dev in enumerate(metadata['Ecephys']['Device']):
-                    if type(dev) is dict:
-                        # Will not overwrite the named device if already in nwbfile
-                        if dev.get('name', defaults['name']) not in nwbfile.devices:
-                            device_kwargs = fill_kwargs_from_defaults(defaults, dev)
-                            nwbfile.create_device(**device_kwargs)
-                    else:
-                        print(f"Warning: Expected metadata['Ecephys']['Device'][{j}] to be"
-                              " a dictionary with keys 'name' and 'description'!"
-                              f" Device [{j}] will not be created.")
-            else:
-                print("Warning: metadata must be a list of dictionaries of the form"
-                      " metadata['Ecephys']['Device'] = [{'name': my_name,"
-                      " 'description': my_description}, ...]!")
+        assert len(metadata.keys()) > 0, "Expected metadata must be a non-empty dictionary!"
+        assert 'Ecephys' in metadata, "Expected metadata to have key 'Ecephys'!"
+        assert 'Device' in metadata['Ecephys'], "Expected metadata['Ecephys'] to have key 'Device'!"
+        assert type(metadata['Ecephys']['Device']) is list, "Expected metadata['Ecephys']['Device'] to be a list!"
+
+        for j, dev in enumerate(metadata['Ecephys']['Device']):
+            assert type(dev) is dict, f"Expected metadata['Ecephys']['Device'][{j}] to be a dictionary!"
+            if dev.get('name', defaults['name']) not in nwbfile.devices:
+                device_kwargs = fill_kwargs_from_defaults(defaults, dev)
+                nwbfile.create_device(**device_kwargs)
```

Comment:
what if metadata was specified as `None`? This will not create any device. Will that be a problem later?

Reply:
In that case, the final lines of

```
 if not nwbfile.devices:
            device_kwargs = fill_kwargs_from_defaults(defaults)
            nwbfile.create_device(**device_kwargs)
```

will trigger, which in the previous format would have always triggered if something went wrong, but in the new format only triggers in the special cases where either metadata was not specified or metadata['Ecephys']['Device'] was empty.

Reply:
All parts of this are now outdated, code block will always create a device with defaults if the appropriate keys are left blank or  metadata is not passed.